### PR TITLE
PP-770: state_count queued jobs count goes negative after server restart

### PR DESCRIPTION
**Pull request by npadole20**
#### Issue-ID

* **[PP-770](https://pbspro.atlassian.net/browse/PP-770)**

#### Problem description
* The value of queued count became negative in the output of qstat -Bf after the server was restarted.

#### Cause / Analysis
* In a server where running and queued jobs are present and when the server is restarted, the count of queued jobs fetched from the datastore becomes 0 inspite of all the jobs being fetched. Thus when these queued jobs are run, the queued count which is now 0 is decremented as expected because of which we get negative values.
#### Solution description
* After server restart, the pbsd_init_reque function is called with either KEEP_STATE or CHANGE_STATE argument. pbsd_init_reque does the task of evaluating the new state and substate and updating these new state. It also internally updated the job state count which was further modified in svr_enquejob function. This led to modifying the global variable maintaining job state count at multiple places after server restart. The fix is to let pbsd_init_reque do the task of only evaluating and updating the job state and substate while letting svr_enquejob take care of updating the job state count once the correct new state and substate is populated.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_qstat_count.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+                                           - counts['Held']
+                                           - counts['Waiting']
+                                           - Running_count)
+        self.logger.info(counts)
+        return counts
+
+    def verify_count(self, Running_count):
+        """
+        Verify that total_jobs and queued_count reported by the output
+        of qstat -Bf matched the number of jobs submitted, where each
```
Comment (prakashcv13) [inline test/tests/functional/pbs_qstat_count.py]: total_jobs will match the number of jobs submitted. queued_count will not.

File: test/tests/functional/pbs_qstat_count.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        self.assertEqual(counts['total_jobs'],
+                         counts['all_state_count'], 'Job count incorrect')
+        self.assertEqual(counts['expected_queued_count'], counts['Queued'],
+                         'Queued count incorrect')
+
+    def test_queued_no_restart(self):
+        """
+        The test case verifies that the reported queued_count in qstat -Bf
+        without a server restart is equal to the number of jobs
+        (each using ncpus=1) submitted.
```
Comment (prakashcv13) [inline test/tests/functional/pbs_qstat_count.py]: the test case verifies that qstat-reported queued_count = total_jobs - running jobs.

File: test/tests/functional/pbs_qstat_count.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        """
+        jid = []
+        for i in range(4):
```
Comment (prakashcv13) [inline test/tests/functional/pbs_qstat_count.py]: a comment above this line to explain the magic number '4', will help.

File: test/tests/functional/pbs_qstat_count.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        a = {ATTR_h: None}
+        j = Job(TEST_USER, a)
+        self.server.submit(j)
+
+        self.submit_waiting_job(600)
+
+        # Wait for jobs to go in R state
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid[0])
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid[1])
+        self.verify_count(2)
```
Comment (prakashcv13) [inline test/tests/functional/pbs_qstat_count.py]: a comment above this line to explain the magic number '2', will help.

File: test/tests/functional/pbs_qstat_count.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        self.server.restart()
+        self.verify_count(2)
+
+    def test_queued_no_restart_multiple_queue(self):
+        """
+        The test case verifies that the queued_count reported in the output
+        of qstat -Bf matches the number of jobs submitted in different queues
+        (each using ncpus=1) without server restart.
+        """
+        # create 2 queues
```
Comment (prakashcv13) [inline test/tests/functional/pbs_qstat_count.py]: create 2 execution queues?

File: test/tests/functional/pbs_qstat_count.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        q1_attr = {ATTR_queue: 'workq1'}
+        q2_attr = {ATTR_queue: 'workq2'}
+
+        # submit jobs, 3 in each queue
+        j = Job(TEST_USER, q1_attr)
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+        j = Job(TEST_USER, q2_attr)
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
```
Comment (prakashcv13) [inline test/tests/functional/pbs_qstat_count.py]: a comment to explain why the submitting of jobs is separated in two parts will help.

File: test/tests/functional/pbs_ralter.py [hunk 7]
```diff
@@ -1273,10 +1273,10 @@ def test_alter_user_permission(self):
...
         rid2, start2, end2 = self.submit_and_confirm_reservation(
             offset, duration, standing=True)
-        new_start1 = self.bu.convert_seconds_to_resvtime(start1 + shift)
-        new_start2 = self.bu.convert_seconds_to_resvtime(start2 + shift)
-        new_end1 = self.bu.convert_seconds_to_resvtime(end1 + shift)
-        new_end2 = self.bu.convert_seconds_to_resvtime(end2 + shift)
+        new_start1 = self.bu.convert_seconds_to_datetime(start1 + shift)
+        new_start2 = self.bu.convert_seconds_to_datetime(start2 + shift)
+        new_end1 = self.bu.convert_seconds_to_datetime(end1 + shift)
+        new_end2 = self.bu.convert_seconds_to_datetime(end2 + shift)
```
Comment (prakashcv13) [inline test/tests/functional/pbs_ralter.py]: thank you for updating pbs_ralter.py

File: test/tests/functional/pbs_qstat_count.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+        TestFunctional.setUp(self)
+        # set ncpus to a known value, 2 here
+        a = {'resources_available.ncpus': 2}
+        self.server.manager(MGR_CMD_SET, NODE, a,
+                            self.mom.shortname, expect=True)
+
+    def submit_waiting_job(self, timedelta):
+        """
+        Submit a job in W state using -a option.
+        The time specified in -a is current time + timedelta.
```
[the remark this answers is not in the document]
Comment (npadole20) [reply test/tests/functional/pbs_qstat_count.py]: I signed the CLA and attached the required PTL logs to the JIRA ticket. I have also addressed the review comments. Please review again.

File: test/tests/functional/pbs_qstat_count.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        # no. of R jobs = 2, as available ncpus=2
+        self.verify_count(2)
+
+    def test_queued_sched_false(self):
+        """
+        This test case verifies that the value of queued_count in the output
+        of qstat -Bf matches the no. of jobs submitted (each using ncpus=1),
+        as scheduling is set to False.
+        """
+        # set scheduling false
```
Comment (bhroam) [inline test/tests/functional/pbs_qstat_count.py]: Comments should explain why, not what.  The line below this says the same thing.  Either explain why you're doing this or remove the comment.

File: test/tests/functional/pbs_qstat_count.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'W'}, id=jid,
+                           offset=30, interval=2)
+
+        jid = self.submit_waiting_job(3)
+        j = Job(TEST_USER)
+        self.server.submit(j)
+        j = Job(TEST_USER)
+        self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'Q'}, id=jid)
```
Comment (bhroam) [inline test/tests/functional/pbs_qstat_count.py]: add an offset=3

File: test/tests/functional/pbs_qstat_count.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        a = {
+            ATTR_stagein: 'inputData@' +
+            self.server.hostname +
+            ':' + os.path.join('noDir', 'nofile')}
+        j = Job(TEST_USER, a)
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'W'}, id=jid,
+                           offset=30, interval=2)
+
+        jid = self.submit_waiting_job(3)
```
Comment (bhroam) [inline test/tests/functional/pbs_qstat_count.py]: Either do an expect here for the job in the 'W' state or add the expect in submit_waiting_job()

File: test/tests/functional/pbs_qstat_count.py [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+            self.server.manager(MGR_CMD_CREATE, QUEUE,
+                                a, qname[i], expect=True)
+
+        q1_attr = {ATTR_queue: 'workq1'}
+        q2_attr = {ATTR_queue: 'workq2'}
+
+        # submit 1 job per queue to ensure a running job in each queue,
+        # then submit 2 more jobs per queue i.e. overall 3 jobs in each queue
+        j = Job(TEST_USER, q1_attr)
+        jid = self.server.submit(j)
```
[the remark this answers is not in the document]
Comment (npadole20) [reply test/tests/functional/pbs_qstat_count.py]: I need jid to check if the job is running. I want one running job in each queue before submitting next jobs.

File: test/tests/functional/pbs_qstat_count.py (hunk 10 above)
Comment (npadole20) [reply test/tests/functional/pbs_qstat_count.py]: I have addressed all your comments. Please review the latest changes.

**Comment by npadole20** [comment]
Hey  @bhroam ,
I have squashed the commits and rebased the branch. 
Please merge it if everything looks in order.

Thanks!

**Comment by bhroam** [comment]
I merged in the PR.

**Pull request merged**